### PR TITLE
Remove uses of advisory 

### PR DESCRIPTION
**Pull request by hannako**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: lib/govspeak/remove_advisory_service.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        "^#{content}^"
+      end
+      if new_body != body
+        record.update!(body: new_body)
+      end
+    end
+
+  private
+
+    def create_minor_update(new_body)
```
Comment (hannako) [inline lib/govspeak/remove_advisory_service.rb:27]: If I understand it correctly, we don't need this method as we are not going to be handling drafts that contain advisory markup anymore - is that right?

If so you could simplify this class, and just initialise it with an `edition`? 
- You could have a method that returns the regex, rather than passing it in as it's not needed elsewhere, only in this class. 
- You could still have the `@body` variable if you did
```
def initialize(object)
  @body = object.body
end

attr_reader :body
```

File: lib/tasks/remove_advisory.rake [hunk 2]
```diff
@@ -1,152 +1,101 @@
...
-      if new_body != body
-        edition.update!(minor_change: true)
-
-        edition.update!(body: new_body)
-
-        puts "Modified body for #{slug}"
-      end
+    published_content_containing_advisory_govspeak.each do |document_id|
+      edition = Document.find(document_id).latest_edition
+      Govspeak::RemoveAdvisoryService.new(document_id, regex).process!
```
Comment (hannako) [inline lib/tasks/remove_advisory.rake:22]: Are you meant to be passing in the edition here, rather than the document_id?

File: lib/govspeak/remove_advisory_service.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  class RemoveAdvisoryService
+    attr_reader :object, :regex, :whodunnit
+
+    def initialize(object, regex, whodunnit = User.find_by(name: "GDS Inside Government Team"))
+      @object = object
+      @regex = regex
+      @whodunnit = whodunnit
+    end
+
+    def process!
```
Comment (hannako) [inline lib/govspeak/remove_advisory_service.rb:11]: I can't figure out how this method could be working. I think `Regexp.last_match(1)` is referencing the `@` rather than the content after the `@`. Should it be `Regexp.last_match(2)` ?

I'm wondering if using the string methods `scan` and `gsub` is making it hard to feel confident about what you're working with, because you have to remember that each subexpression in a `()` in your regexp equates to an object in the array that scan generates. If you'd used the Regexp method `match` it outputs the matches as a hash that you can [reference by key](https://ruby-doc.org/3.3.6/Regexp.html#class-Regexp-label-Named+Captures). The downside would be it only matches once, rather than searching the entire body, so for content with more than one advisory you'd need to think about recursion for example.

TLDR: 
- If you've had this code working with `scan` and `gsub` then stick with it, but please can you add comprehensive tests to show that your regex will match the different types of advisory markup that exists (eg @ only at the start, @ at the start and the end etc), and replace it as expected. 

- If you're struggling to get this to work, I wrote it using `Regexp.match` just to understand the shape of the problem, and I think it works so I could share that with you to see what you think...but I don't want to de-rail if you're almost there! So just let me know.

File: lib/govspeak/remove_advisory_service.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      body = object.body
+      matches = body.scan(regex)
+      return if matches.empty?
+
+      new_body = body.gsub(/(^@)([\s\S]*?)(@?)(?=(?:^\$CTA|\r?\n\r?\n|^@|$))/m) do
```
Comment (hannako) [inline lib/govspeak/remove_advisory_service.rb:16]: I've had a go and splitting your regex:

```
def regexp_for_advisory_markup
  opening_at = '(^@)'
  content_after_at = '([\s\S]*?)'
  closing_at = '(@?)'
  other_possible_line_ends = '(?:^\$CTA|\r?\n\r?\n|^@|$)'
  Regexp.new(opening_at + content_after_at + closing_at + '(?=' + 
  other_possible_line_ends + ')', Regexp::MULTILINE)
end
```

**Review by hannako — COMMENTED** [review]
I opened your branch as a draft PR so it was easier to review! Hope that's ok.

Great work on figuring out that regex! So mind melting, but think I got there in the end and had a go and splitting it out - see what you think. 

I've left some comments - happy to chat through or pair, whatever is best for you.

File: db/data_migration/20241024103203_remove_advisory_govspeak_from_edition_body.rb [hunk 5]
```diff
@@ -0,0 +1,142 @@
+CONTENT_CONTAINING_ADVISORY_GOVSPEAK = %w[ae74f681-0eac-4de0-bedc-b4c1703ba5f5
```
[the remark this answers is not in the document]
Comment (beccapearce) [reply db/data_migration/20241024103203_remove_advisory_govspeak_from_edition_body.rb]: Yes this will be removed!

File: lib/govspeak/remove_advisory_service.rb (hunk 1 above)
Comment (beccapearce) [reply lib/govspeak/remove_advisory_service.rb:27]: I think it would be good to keep this in, as it will match our flow for normal editions. So the changes will appear in the audit history for the doc

File: lib/govspeak/remove_advisory_service.rb (hunk 3 above)
Comment (beccapearce) [reply lib/govspeak/remove_advisory_service.rb:11]: YES PLEASE! A solution with `.match` would be so much better! I think this was the issue I was trying to fix in my tests! <3

File: lib/govspeak/remove_advisory_service.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+          # Find the attachment in the new draft
+          new_attachment = draft.html_attachments.find_by(slug: @object.slug)
+
+          # Replace advisories in the body of the new attachment
+          new_body = replace_all_advisories(new_attachment.body)
+          new_attachment.govspeak_content.update!(body: new_body)
+
+          # Update the draft edition and publish
+          draft.update!(minor_change: true)
+          draft.submit!
```
Comment (hannako) [inline lib/govspeak/remove_advisory_service.rb:45]: You could probably DRY up the `process!` method by creating a ` submit_and_publish(draft)`

File: lib/tasks/remove_advisory.rake [hunk 7]
```diff
@@ -0,0 +1,68 @@
+namespace :remove_advisory_from_editions do
+  desc "Process advisory govspeak in published editions"
+  task published_editions: :environment do
+    regex = Regexp.new(/(^@)([\s\S]*?)(@?)(?=(?:^\$CTA|\r?\n\r?\n|^@|$))/m).to_s
```
Comment (hannako) [inline lib/tasks/remove_advisory.rake:4]: There's an `EmbeddedContentPatterns` module in Whitehall that contains regex's used by the various extractor classes in lib/Govspeak. I think we should follow the same pattern here rather than defining the regexp in two files.

i.e. create a `Govspeak::AdvisoryPattern` module and include it in both your new service and your rake task

**Review by ChrisBAshton — COMMENTED** [review]
Looks good in principle 👍 some comments below.

**Comment by beccapearce** [comment]
Thank you Chris! I've added the changes as a separate commit for now, just to make it easy to rereview. I'll squash them together before we get ready to merge.

**Review by catalinailie — APPROVED** [review]

**Comment by hannako** [comment]
I don't think I can approve this PR Becca, as I opened it on your draft branch when you were off. But it looks great to me!
